### ISSUE TITLE
how perform transaction to smart contract?

### ISSUE THREAD

**Issue by San4es-hue** [pink-room/walletconnectkit-android#11]
Hi, How can I perform transaction to smart contract? Thanks

**Comment by Neo-glitch** [comment]
Hi, how can I interact with a smart contract? Thanks in advance

**Comment by bffcorreia** [comment]
Hi all! To perform a transaction to a smart contract you need to use the `performTransaction` method and pass the `data` parameter. This parameter should be the encoded function call of the smart contract. To encode the function you need to use other tools like [web3j](https://docs.web3j.io/4.8.7/).

I want to write a blog post explaining this but I'm currently limited on time. So, it might take some time before I share it publicly.

**Comment by mariavieira99** [comment]
What if you don't want to pass anything in the value parameter of performTransaction? I've already tried to pass the value "0.0", it opens my wallet but no confirmation popup appears

**Comment by f4cker** [comment]
@San4es-hue Hope the cheat code would be helpful for you.

![codeshot 2022-04-21 at 17 00 03 PM](https://user-images.githubusercontent.com/3525337/164419881-13195fd3-f50e-4e24-9f8f-774026772e13.png)

**Comment by jayvs6341** [comment]
@f4cker 
What is **tokenUri** ?

Can you please provide more details about the code

**Comment by f4cker** [comment]
@jayvs6341 
As the code shows that the first list is a collection of input param, and the second list is a collection of output param, and all of this are according to functions declared in your contract

**Comment by jayvs6341** [comment]
@f4cker 

Can you please explian me,

**Utf8String(tokenUri)**

Here is what is **tokenUri** ?

Here is my code.

          val function = Function("transfer", listOf(Address(walletConnectKit.address),
                Utf8String("file:///android_asset/Token.json")
            ),
                emptyList<TypeReference<*>>() )

            val data = FunctionEncoder.encode(function)

            val web3j = Web3j.build(HttpService("https://polygon-mumbai.infura.io/v3/f7587fb94ddb497c90cb7edb88ad3395"))

            lifecycleScope.launch {
                withContext(Dispatchers.IO){
                    val nounce = web3j.ethGetTransactionCount(walletConnectKit.address,DefaultBlockParameterName.LATEST).send().transactionCount
                    kotlin.runCatching {
                        walletConnectKit.performTransaction("0xa8E59566Ab7c6FD5729f61388131099ce02d074E",value = "", data , nounce.toString())
                    }
                }.onSuccess {
                    Log.e("OnSuccess",":-"+it.result)
                }.onFailure {
                    Log.e("onFailure",":-"+it.message)
                }
            }

I am getting following error.

**Bad offset/length: offset=0 len=0 in.length=0**

Can you please help me to sort it out, I am badly stuck at this.

**Comment by f4cker** [comment]
@jayvs6341 
There is no matter about the 'tokenUri', may be you should make the 'value' param send a "0" in 'performTransaction' method which I have refactored it and try it again.
<img width="664" alt="image" src="https://user-images.githubusercontent.com/3525337/164586029-50382d2c-7946-4b08-9865-3bc622bb15e7.png">

**Comment by jayvs6341** [comment]
@f4cker 

error resolved.

But when it redirects to Metamask, it dont show the transaction approval dialog.

I have my own smart contract tokens. And I imported that tokens to polygon test net.

I have 100 tokens in my metamask and I want to send 10 to other address using my application.

If I try the default performTranscation function, it always sends the default test net currency. Like ether, matics etc.

How can I initiate transaction to send "MyToken" ?

**Comment by f4cker** [comment]
@jayvs6341 
That should be about the business, may be you should invoke the **approve** method in contract first to make sure your contract has the privilege to transfer tokens

**Comment by jayvs6341** [comment]
@f4cker 

yes, I have the function "transfer" to transfer the tokens, and I want metamask to approve the transaction.

If we go to metamask, we can manylly send the tokens correct,

Same think I want to invoke through Dapp, so it will redirect to metamask and then user just needs to approve it.

**Comment by f4cker** [comment]
@jayvs6341 
You didn't get it,  the **approve** method is just a transaction like **transfer**, you should invoke it with **performTransaction** to interact with the smart contract through MetaMask, and then the MetaMask would show the approval dialog for you to confirm the transaction.

**Comment by jayvs6341** [comment]
@f4cker 

Thanks a lot for your ongoing support.
Yeah, trying to understand

It would be very helpful, if you can provide me some thing like sample code

**Comment by mazrou** [comment]
Hello @jayvs6341,
I tried your code . but I get and error on building : 

`Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing com.android.build.gradle.internal.tasks.CheckDuplicatesRunnable`

note that I import the methods that you use from  
` implementation 'org.web3j:codegen:5.0.0'`
and I'm using :
`implementation 'dev.pinkroom:walletconnectkit:0.3.2'`

please if you can help me and give us more details about your code if it works

**Comment by Ravikumarsym** [comment]
Hi @San4es-hue 
gas price sent to metamask transaction popup does not displayed

gas price null to invalid amount popup displayed

**Comment by anishsuwal** [comment]
Did any one solve the problem. I been trying to implement the same but no luck. Can any one help.

**Comment by joeleonjr** [comment]
Hi! I'm a security researcher @TruffleSecurity. I found a valid API key leaked in a comment in this Issue. I highly recommend rotating the API key immediately (not just deleting the comment). Here's a blog post I wrote on remediating these types of vulnerabilities: https://trufflesecurity.com/blog/remediate-leaked-api-keys-with-key-rotation.<br><br>If you can't find the leaked API key, you can use the following [TruffleHog](https://github.com/trufflesecurity/trufflehog) command to automatically scan for it:<br>`trufflehog github --repo https://github.com/<your_username>/<your_reponame> --issue-comments --only-verified`<br><br>If it's not clear where/what the security issue is, please let me know. I'd be happy to help. Please note that the API key might be in the past edit of a comment.<br><br><br><br>(cc @jayvs6341)
